py/torch_tensorrt/dynamo/conversion/impl/elementwise/base.py

```diff
@@ -139,27 +143,51 @@ def convert_binary_elementwise(
 
     if trt_promoted_type != lhs_val.dtype:
         lhs_val = cast_trt_tensor(
-            ctx, lhs_val, trt_promoted_type, name, target, source_ir
+            ctx, lhs_val, trt_promoted_type, f"{name}_cast_lhs_val", target, source_ir
         )
     if trt_promoted_type != rhs_val.dtype:
         rhs_val = cast_trt_tensor(
-            ctx, rhs_val, trt_promoted_type, name, target, source_ir
+            ctx, rhs_val, trt_promoted_type, f"{name}_cast_rhs_val", target, source_ir
         )
 
-    # Check the limitation in the doc string.
-    if ctx.net.has_implicit_batch_dimension:
-        if is_lhs_trt_tensor and not is_rhs_trt_tensor:
-            assert len(lhs_val.shape) >= len(
-                rhs_val.shape
-            ), f"{lhs_val.shape} >= {rhs_val.shape}"
-        elif not is_lhs_trt_tensor and is_rhs_trt_tensor:
-            assert len(rhs_val.shape) >= len(
-                lhs_val.shape
-            ), f"{rhs_val.shape} >= {lhs_val.shape}"
-
-    lhs_val, rhs_val = broadcast(
-        ctx.net, lhs_val, rhs_val, f"{name}_lhs", f"{name}_rhs"
-    )
+    if has_dynamic_shape(lhs_val.shape) or has_dynamic_shape(rhs_val.shape):
+        lhs_val, rhs_val = broadcast(
+            ctx.net, lhs_val, rhs_val, f"{name}_lhs", f"{name}_rhs"
+        )
+    else:
+        lhs_val_shape = lhs_val.shape
```

Comment:
@peri044 

I followed the code changes to address the error caused by broadcasting, and it seems to be related to this issue.

If we perform element-wise operations on two inputs with shapes lhs `(2, 2, 14, 1, 5)` and rhs `(2, 1, 3, 5)`, the output of the element-wise operation layer should be `(2, 2, 14, 3, 5)`. 
However, with the current code, the output of the element-wise operation layer follows the larger rank of lhs_val and becomes `(2, 2, 14, 1, 5)`. 

As a result, an error that did not occur on the local machine was found during CI/CD (https://github.com/pytorch/TensorRT/pull/2726)

Is it necessary to make adjustments to fix this code?

Reply:
Yes, it is necessary to fix this code as any converter that uses this would likely have that failure. It sounds like the fix would be broadcasting across mismatched dimensions instead of using rank heuristic. @zewenli98 probably you have an idea about these specific changes while fixing test_binary_ops ?